### DISCUSSION TITLE
Inline Styles

### DISCUSSION THREAD

**Discussion by ainsleyclark** [a-h/templ#762] [General]
Hi,

I'm struggling to understand inline styles.
I would like to apply a width property aka `style="width: 50%"

```
type FormGroupProps struct {
	Width       *int
}

func width(percent *int) string {
	return fmt.Sprintf("%d%%", percent)
}

templ FormGroup(props FormGroupProps) {
	<fieldset
		if props.Width != nil {
			style={ templ.SafeCSSProperty(width(props.Width)) }
		}
	>
}
```

I'm unsure why the docs indicate I need to use class names for styling?

**Comment by joerdav** [comment]
As of now yes, the correct way to do what you are trying is:

```
type FormGroupProps struct {
	Width       *int
}

css width(percent *int) {
	width: { fmt.Sprintf("%d%%", percent) };
}

templ FormGroup(props FormGroupProps) {
	<fieldset
		if props.Width != nil {
			class={ width(props.Width) }
		}
	>
}
```

And this can be taken further with `KV`:

```
type FormGroupProps struct {
	Width       *int
}

css width(percent *int) {
	width: { fmt.Sprintf("%d%%", percent) };
}

templ FormGroup(props FormGroupProps) {
	<fieldset
		class={ templ.KV(width(props.Width), props.Width != nil) }
	>
}
```

**Reply by ainsleyclark** [reply]
Thanks for your explanation @joerdav 

Is there a reason why this cannot be done in the `style` attribute?

**Reply by joerdav** [reply]
It was originally a security concern, however we are reconsidering it, we may open up the style attr so your original solution would work.

**Reply by JWTappert** [reply]
Do the css components allow params? I am trying to pass in a couple params similar to your examples and get parse errors.

**Reply by ainsleyclark** [reply]
@joerdav do you know when this will be likely be incorporated into prod?

**Reply by joerdav** [reply]
@JWTappert Yes they do support params, they must be strings: https://templ.guide/syntax-and-usage/css-style-management#css-component-arguments

**Reply by joerdav** [reply]
@ainsleyclark there currently hasn't been any work done to relax the rules around style attributes. The wider css discussion is happening here: https://github.com/a-h/templ/issues/740

But I believe that a PR to accept strings in the style tag would be accepted. It would involve removing this code and any tests that lock ensure it: https://github.com/a-h/templ/blob/485465e823b8c4e249b6067276631d98ff4e62fe/parser/v2/types.go#L489-L496

And updating the docs to remove mentions of it: https://github.com/a-h/templ/blob/485465e823b8c4e249b6067276631d98ff4e62fe/docs/docs/10-security/01-injection-attacks.md?plain=1#L34

**Comment by JWTappert** [comment]
These examples are extremely helpful. Commenting so I can follow updates as I am trying to do something similar.

**Comment by eslam-allam** [comment]
I don't quite understand how the css components are supposed to work though. I understand that templ will generate a <style> element above the element that used the css component with a random class name. Every time i attempt to use it though the <style> element gets converted to a string and doesn't work. For example:

this template:

```templ
css stashBackground(stash model.UserStash) {
    background-color: {"#" + stash.ColourHex};
  }

templ StashList(stashes []model.UserStash, options model.StashOverviewItemsOptions) {
	<select
		name="stashId"
		id="userStashes"
		class="stash-item-criteria select select-bordered w-full max-w-xs h-full disabled:skeleton disabled:z-0"
		if len(stashes) == 0 {
			disabled
		}
		hx-get="/user/stashes/items"
		hx-target="#stashItems"
		hx-swap="outerHTML"
		hx-params="*"
		hx-indicator="#itemsSpinner"
	>
		<option disabled selected>Select a Stash Tab</option>
		for _, stash := range stashes {
			<option class={ stashBackground(stash) } value={ stash.Id }>{ stash.Name }</option>
		}
	</select>
}
```

becomes this when generated:
```html
<select name="stashId" id="userStashes" class="stash-item-criteria select select-bordered w-full max-w-xs h-full disabled:skeleton disabled:z-0" hx-get="/user/stashes/items" hx-target="#stashItems" hx-swap="outerHTML" hx-params="*" hx-indicator="#itemsSpinner">
                        <option disabled selected>Select a Stash Tab</option>
                        <style type="text/css">
                            .stashBackground_43c3 {
                                background-color: #ffffff;
                            }
                        </style>
                        <option class="stashBackground_43c3" value="">Test</option>
                    </select>
```
then the browser places it in the DOM like this:

```html
<select name="stashId" id="userStashes" class="stash-item-criteria select select-bordered w-full max-w-xs h-full disabled:skeleton disabled:z-0" hx-get="/user/stashes/items" hx-target="#stashItems" hx-swap="outerHTML" hx-params="*" hx-indicator="#itemsSpinner">
  <option disabled="" selected="">Select a Stash Tab</option>
  .stashBackground_43c3{background-color:#ffffff;}
  <option class="stashBackground_43c3" value="" style="background-color: #ffffff;">Test</option>
</select>
```
Then the styles dont get applied,
I'm probably using this incorrectly but I can't find any reference to an actual working use of dynamic css using templ.
